doc/mgmt/gnmi/save_on_set.md

```diff
@@ -0,0 +1,193 @@
+  - [sonic-gnmi](https://github.com/sonic-net/sonic-gnmi)
+- Module/sub-module interfaces and dependencies.
+  - Adds a flag to the gNMI module.
+- SWSS and Syncd changes in detail
+  - N/A
+- DB and Schema changes (APP_DB, ASIC_DB, COUNTERS_DB, LOGLEVEL_DB, CONFIG_DB, STATE_DB)
+  - N/A
+- Sequence diagram if required.
+![Flow Char](images/save_on_set.png)
+- Linux dependencies and interface
+  - N/A
+- Warm reboot requirements/dependencies
+  - None
+- Fastboot requirements/dependencies
+  - None
+- Scalability and performance requirements/impact
+  - If the feature is enabled, there is performance hit that scales with the size of ConfigDB.
+- Memory requirements
+  - Negligible, a couple function pointers.
+- Docker dependency
+  - None
+- Build dependency if any
+  - None
+- Management interfaces - SNMP, CLI, RestAPI, etc.,
+  - [Configuration and management](#configuration-and-management)
+- Serviceability and Debug (logging, counters, trace etc) related design
+  - Logging of feature being enabled;
+- Is this change specific to any platform? Are there dependencies for platforms to implement anything to make this feature work? If yes, explain in detail and inform community in advance.
+  - Any Platform
+- SAI API requirements, CLI requirements, ConfigDB requirements. Design is covered in following sections.
+
+#### Required Changes
+
+##### Telemetry Executable
+
+A new command-line parameter will be added to control the behavior of the save-on-set functionality. By default, i.e. when the option is not specified, the gNMI server will behave as it did before this change - the configuration will not be saved to a file without explicit action from the administrator for example by execution of a command via ssh connection.
+
+The telemetry.sh startup script will be modified to read `TELEMETRY|gnmi|save_on_set` and pass the value along during startup. Any changes to this value will only take effect after restarting telemetry.
+
+The new parameter will be: --with-save-on-set and when present it will configure a function pointer variable gnmi.SaveOnSet to point to a function that initiates the save operation.
+
+```go
+var (
+  withSaveOnSet = flag.Bool("with-save-on-set", false, "Enables save-on-set.")
+)
+
+// ...
+
+if *withSaveOnSet {
+  gnmi.SaveOnSet = gnmi.SaveOnSetEnabled
+}
+```
+
+##### gNMI.Set() Handler
+
+The RPC will execute the save-on-set function at the end of each call.
+
+```go
+// SaveOnSetEnabled saves configuration to a file
+func SaveOnSetEnabled() {
+  transformer.SaveStartupConfig()
+}
+// SaveOnSetDisabeld does nothing.
+func SaveOnSetDisabled() {}
+
+// SaveOnSet point to a function that is called to save changes of configuration
+// to a file. By default it points to an empty function - the configuration is not
+// saved to a file.
+var SaveOnSet = SaveOnSetDisabled
+
+func (s *Server) Set(ctx context.Context, req *gnmipb.SetRequest) (*gnmipb.SetResponse, error) {
+
+  // ...
+
+  SaveOnSet()
```

Comment:
If set operations failed here, do you still need SaveOnSet?

Reply:
The call to save on set is the last thing before returning. If there are errors set failures the function returns the err early and the save is never called.

Looking over the code I do see one path were a failure is possible and the code doesn't return early. We could make this an explicit check if its necessary.

Reply:
https://github.com/sonic-net/sonic-gnmi/pull/242